### PR TITLE
ShadowDOM Support: fix active element receive and caret position calculation

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: js/core/dom_adapter.js [hunk 1]
```diff
@@ -121,8 +121,10 @@ const nativeDOMAdapterStrategy = {
         return this._document;
     },
 
-    getActiveElement() {
-        return this._document.activeElement;
+    getActiveElement(element) {
+        const doc = element?.getRootNode?.() ?? this._document;
```
Comment (ksercs) [inline js/core/dom_adapter.js]:
```suggestion
        const activeElementHolder = element?.getRootNode?.() ?? this._document;
```

File: testing/helpers/shadowDOM.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+function appendShadowRoot(shadowSelector) {
```
Comment (ksercs) [inline testing/helpers/shadowDOM.js]: According to style guides we write abbreviations in CamelCase.
`testing/helpers/shadowDom.js`

File: testing/helpers/shadowDOM.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    root.attachShadow({ mode: 'open' });
+    root.shadowRoot.appendChild(container);
+    container.appendChild(control);
+
+    this.root = root;
+    this.container = container;
+    this.control = control;
+}
+
+exports.appendShadowRoot = appendShadowRoot;
```
Comment (ksercs) [inline testing/helpers/shadowDOM.js]: why not `export { appendShadowRoot }`?

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js [hunk 4]
```diff
@@ -2231,3 +2232,36 @@ QUnit.module('valueChanged should receive correct event parameter', {
...
+
+        const instance = $numberBox.dxNumberBox('instance');
+        const $numberBoxInput = $numberBox.find(`.${INPUT_CLASS}`);
+        const mouse = pointerMock($numberBoxInput);
+
+        $numberBoxInput.focus();
+
+        mouse.wheel(10);
+
+        assert.equal(instance.option('value'), 101.6, 'value is greeter after mousewheel up');
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]: we try to use `strictEqual` everywhere

Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]:
```suggestion
        assert.equal(instance.option('value'), 101.6, 'value is increased after mousewheel up');
```

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js [hunk 5]
```diff
@@ -2231,3 +2232,36 @@ QUnit.module('valueChanged should receive correct event parameter', {
...
+
+        mouse.wheel(-20);
+        assert.equal(instance.option('value'), 100.6, 'value is less after mousewheel down');
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]:
```suggestion
        assert.strictEqual(instance.option('value'), 100.6, 'value is decreased after mousewheel down');
```

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js [hunk 6]
```diff
@@ -2231,3 +2232,36 @@ QUnit.module('valueChanged should receive correct event parameter', {
...
+
+        $numberBoxInput.blur();
+        mouse.wheel(-20);
+        assert.roughEqual(instance.option('value'), 100.6, 1.001);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]: if u use just `100` as the numberbox value you will not need to use `roughEqual`

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js [hunk 7]
```diff
@@ -2231,3 +2232,36 @@ QUnit.module('valueChanged should receive correct event parameter', {
...
+        $(this.container).empty();
+    },
+}, () => {
+    QUnit.test('should change value on mouse wheel', function(assert) {
+        const $numberBox = $(this.control).dxNumberBox({
+            value: 100.6
+        });
+
+        const instance = $numberBox.dxNumberBox('instance');
+        const $numberBoxInput = $numberBox.find(`.${INPUT_CLASS}`);
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]:
```suggestion
        const $input = $numberBox.find(`.${INPUT_CLASS}`);
```

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js [hunk 8]
```diff
@@ -2230,3 +2231,41 @@ QUnit.module('symbol with dot in format', {
...
     });
 });
+
+QUnit.module('ShadowDOM', {
+    beforeEach: function() {
+        appendShadowRoot.call(this, '#numberbox');
+    }
+}, function() {
+    QUnit.test('should move caret', function(assert) {
+        this.$element = $(this.control).dxNumberBox({
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/mask.tests.js]: seems like `this` using inside of test body is excess

File: testing/helpers/shadowDOM.js (hunk 3 above)
Comment (AlexKamaev) [reply testing/helpers/shadowDOM.js]: Probably just a mistake

File: testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js (hunk 6 above)
Comment (AlexKamaev) [reply testing/tests/DevExpress.ui.widgets.editors/numberBoxParts/common.tests.js]: I copied this test and adjusted it to work inside that ShadowDOM. I think we can easily remove this `roughEqual` assertion cause we do not need to do some extra checks

**Comment by AlexKamaev** [comment]
Ready for review

**Review by nightskylark — COMMENTED** [review]
LGTM

**Comment by AlexKamaev** [comment]
@ksercs

**Review by ksercs — APPROVED** [review]

**Pull request merged**